src/rtutils.c

```diff
@@ -526,11 +528,44 @@ static size_t jl_static_show_x_(JL_STREAM *out, jl_value_t *v, jl_datatype_t *vt
     }
     else if (vt == jl_datatype_type) {
         jl_datatype_t *dv = (jl_datatype_t*)v;
-        if (dv->name->module != jl_core_module) {
+        jl_sym_t *globname = dv->name->mt != NULL ? dv->name->mt->name : NULL;
+        int globfunc = globname && jl_get_global(dv->name->module, globname) &&
+            jl_typeof(jl_get_global(dv->name->module, globname)) == v &&
+            !strchr(jl_symbol_name(globname),'#');
+        jl_sym_t *sym = globfunc ? globname : dv->name->name;
+        char *sn = jl_symbol_name(sym);
+        int hidden = !globfunc && strchr(sn,'#');
+        size_t i = 0;
+        int quote = 0;
+        if (hidden) {
+            n += jl_printf(out, "getfield(");
+        }
+        else if (globfunc) {
+            n += jl_printf(out, "typeof(");
+        }
+        if (dv->name->module != jl_core_module || !jl_module_exports_p(jl_core_module, sym)) {
```

Comment:
`jl_module_exports_p` can allocate and can't be used here (or should be fixed not to allocate?)